File: day05/solution.jl
COLUMN_WIDTH = 4

function readStartMap(file)
  buffer = []
  line = readline(file)
  columnCount = getColumnCount(line)
  
  while !eof(file) && columnCount < 0
    push!(buffer, line)
    line = readline(file)
    columnCount = getColumnCount(line)
  end

  startMap = fill("", columnCount)
  for content in buffer
    for i = 1:columnCount
      crate = content[(i - 1)*COLUMN_WIDTH+2]
      if crate != ' ' 
        startMap[i] = string(crate, startMap[i])
      end
    end
  end

  return startMap
end

function getColumnCount(content::String)
  fields = split(content)
  if all(isNumber(string(f)) for f in fields)
    return length(fields)
  end
  return -1
end

function isNumber(content::String)
  try 
    parse(Int, content)
  catch ArgumentError
    return false
  end
  return true
end

function apply!(stacks, n, src, dest, order)
  stacks[dest] = string(stacks[dest], order(string(stacks[src][end-n+1:end])))
  if length(stacks[src]) == n 
    stacks[src] = ""
  else
    stacks[src] = stacks[src][1:end-n]
  end
end

open("input", "r") do file
  crane1 = readStartMap(file)
  crane2 = copy(crane1)

  f(x) = x
  while !eof(file)
    instruction = split(readline(file))
    if length(instruction) > 0
      n = parse(Int, instruction[2])
      src = parse(Int, instruction[4])
      dest = parse(Int, instruction[6])

      apply!(crane1, n, src, dest, reverse)
      apply!(crane2, n, src, dest, f)
    end
  end

  println("Crane 9000: $crane1")
  println("Crane 9001: $crane2")
end;

File: day05/solution.go
package main

import (
	"bufio"
	"fmt"
	"log"
	"os"
	"strconv"
	"strings"
)

const COLUMN_WIDTH = 4

type Stacks []string

func check(err error) {
	if err != nil {
		log.Fatalf("Error: %v\n", err)
	}
}

func readStartMap(scanner *bufio.Scanner) Stacks {
	scanner.Scan()
	line := scanner.Text()
	buffer := make([]string, 0)

	columnCount := countColumns(line)
	// This parsing puts the last element at the front of the slice
	for columnCount < 0 {
		buffer = append(buffer, line)
		if !scanner.Scan() {
			log.Fatalln("out of content")
		}
		line = scanner.Text()
		columnCount = countColumns(line)
	}

	startMap := make([]string, columnCount)
	for _, content := range buffer {
		for i := 0; i < columnCount; i++ {
			crate := content[i*COLUMN_WIDTH+1]
			if crate != ' ' {
				startMap[i] = startMap[i] + string(crate)
			}
		}
	}

	return Stacks(startMap)
}

func countColumns(line string) int {
	fields := strings.Fields(line)
	for i, token := range fields {
		value, err := strconv.Atoi(token)
		if err != nil || value != i+1 {
			return -1
		}
	}
	return len(fields)
}

func reverse(content string) string {
	result := []byte(content)
	for i, j := 0, len(result)-1; i < j; i, j = i+1, j-1 {
		result[i], result[j] = result[j], result[i]
	}
	return string(result)
}

func (stacks Stacks) move(n, src, dest int, fn func(string) string) {
	stacks[dest] = fn(stacks[src][0:n]) + stacks[dest]
	stacks[src] = stacks[src][n:]
}

func main() {
	file, err := os.Open("input")
	check(err)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	crane1 := readStartMap(scanner)
	crane2 := make(Stacks, len(crane1))
	copy(crane2, crane1)

	fmt.Printf("Starting stack is: %v\n", crane1)

	// Find the the first instruction
	scanner.Scan()

	// Parse the instructions
	for scanner.Scan() {
		line := scanner.Text()

		var n, src, dest int = 0, 0, 0

		read, err := fmt.Sscanf(line, "move %d from %d to %d", &n, &src, &dest)
		check(err)
		if read < 3 {
			log.Fatalf("Expected 3 items to be parsed, got %d\n", read)
		}

		crane1.move(n, src-1, dest-1, reverse)
		crane2.move(n, src-1, dest-1, func(x string) string { return x })
	}
	fmt.Printf("Crane 9000: %v\n", crane1)
	fmt.Printf("Crane 9001: %v\n", crane2)
}
